src/sdk/wallet.js

```diff
@@ -0,0 +1,192 @@
+        const amountWei = ethers.utils.parseEther(amountEth.toString());
+        const clientFund = _clientFund.get(this);
+        dbg('Sending ' + amountWei + ' wei to ' + clientFund.address + ' from ' + this.address);
+        const tx = await this.send(clientFund.address, amountWei, options);
+        return getTransactionReceipt.call(this, tx.hash);
+    }
+
+    /**
+     * Deposits a token from the on-chain balance of the wallet to striim.
+     * @see https://docs.ethers.io/ethers.js/html/api-providers.html#transaction-receipts
+     * @param {(number|string)} amount The amount of currency to deposit.
+     * @param {string} symbol The currency symbol
+     * @param [options]
+     * @returns {Promise<Object[]>} A promise that resolves into an array of transaction receipts.
+     */
+    async depositToken(amount, symbol, options) {
+        const tokenInfo = await getTokenInfo.call(this, symbol);
+        const amountBN = ethers.utils.parseUnits(amount.toString(), tokenInfo.decimals);
+        const tokenContract = new Erc20Contract(tokenInfo.currency, this);
+        const clientFund = _clientFund.get(this);
+
+        try {
+            var approvalTx = await tokenContract.approve(clientFund.address, amountBN, options);
+            dbg('ApproveTX: ' + JSON.stringify(approvalTx));
+
+            var approvalTxReceipt = await getTransactionReceipt.call(this, approvalTx.hash);
+            dbg('ApproveTX (receipt): ' + JSON.stringify(approvalTxReceipt));
+        }
+        catch (err) {
+            dbg(err);
+            throw new Error('Failed to approve ERC20 token payment in time!')
+        }
+
+        if (approvalTxReceipt.status === 0)
+            throw new Error('Pre-approval transaction failed!');
+
+        try {
+            var depositTx = await clientFund.depositTokens(tokenContract.address, amountBN, options);
+            dbg('DepositTX: ' + JSON.stringify(depositTx));
+
+            var depositTxReceipt = await getTransactionReceipt.call(this, depositTx.hash);
+            dbg('DepositTX (receipt): ' + JSON.stringify(depositTxReceipt));
+        }
+        catch (err) {
+            dbg(err);
+            throw new Error('Failed to deposit token in time!')
+        }
+
+        if (depositTxReceipt.status === 0)
+            throw new Error('Deposit transaction failed!');
+
+        return [approvalTxReceipt, depositTxReceipt];
+    }
+}
+
+module.exports = Wallet;
+
+/**
+ * Private method - invoke bound to instance.
+ * Attempts to get transaction receipt once every second for up to 1 minute.
+ * @see https://docs.ethers.io/ethers.js/html/api-providers.html#transaction-receipts
+ * @param {string} transactionHash
+ * @returns {Promise<Object>}
+ */
+function getTransactionReceipt(transactionHash) {
+    let retries = 0;
+    return Promise.retry(() => {
+        process.stderr.write('.');
+        return deferredIsNotNull(this.provider.getTransactionReceipt(transactionHash))
```

Comment:
Does the implementation of getTransactionReceipt return error on 404? If not, this will not work I think, since it will not retry since the catch handler of the retry was not invoked.

Some HTTP client libraries don't report error on 404, while others do, and sometimes it's configurable.

Reply:
AFAIK superagent reports all errors.